### PR TITLE
Refactor kton precompile

### LINKED ISSUE #575: KTON_PROCOMPILE issues
Function `transfer_and_call(address wkton_addr_,uint256 value_)`

### Function selector params issue.
`transfer_and_call(address,address,uint256)`  ==> `transfer_and_call(address,uint256)`

The function only use two paramters, but the sig is three.

### Transfer decimal issue.
`transfer_and_call(address,uint256)`
The second parameter `value_` decimal should be the same as `PKTON`, maybe `9`

When I transfer 1 `PKTON`, the `value_` should be `1 * 1e9`, but now is `1`.

### Transfer succeed when `wkton_addr_` is KTON_PROCOMPILE itself or any other address.
When do `wkton_addr_.deposit()`. If the call is failed, the transaction should revert. 

#### Script
```shell
curl -fsS https://pangolin-rpc.darwinia.network -d '{"id":1,"jsonrpc":"2.0","method":"eth_estimateGas","params":[{"data":"0x784deab50000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000000a","from":"0x0f14341A7f464320319025540E8Fe48Ad0fe5aec","gas":"0x5b8d80","to":"0x0000000000000000000000000000000000000016"}]}' -H 'Content-Type: application/json'
```
#### Node info
```shell
2021-04-02 15:12:50.486 DEBUG        http.worker00 evm: Execution Error(Other("Invalid Action！")) [source: 0x0000000000000000000000000000000000000016, value: 0, gas_limit: 999978656, actual_fee: 0]
2021-04-02 15:12:50.487 DEBUG        http.worker00 evm: Execution Succeed(Returned) [source: 0x0f14341a7f464320319025540e8fe48ad0fe5aec, value: 0, gas_limit: 1000000000, actual_fee: 0]
2021-04-02 15:12:50.487 DEBUG        http.worker00 evm: Execution Error(Other("Invalid Action！")) [source: 0x0000000000000000000000000000000000000016, value: 0, gas_limit: 499989156, actual_fee: 0]
2021-04-02 15:12:50.487 DEBUG        http.worker00 evm: Execution Succeed(Returned) [source: 0x0f14341a7f464320319025540e8fe48ad0fe5aec, value: 0, gas_limit: 500010500, actual_fee: 0]
```

### PR DESCRIPTION
**Pull request by AsceticBear**
This pr fix #575 #578 and implemented based on #577. Although this pr must wait for #577 merged. It is ready for reviewing now.

This precompile is **critical** to the whole system, so please make sure you spend some time reviewing it.

@HackFisher @hujw77 @AurevoirXavier 

---

With regard to the design of the KTON precompile, I think it necessary to write the following material to help those who are not familiar with this precompile contract to understand it.

Precompile contracts are special contracts that have a fixed address and the user can send transactions to call methods in the contract. The KTON contract we are discussing is a pre-compiled contract with a fixed address of `0000000000000000000000000000000000000016`.

The purpose of the KTON precompile is to support the transfer of the native token KTON across the Darwinia network and the DVM warped contract like [WCKTON Contract](https://github.com/evolutionlandorg/evo-deploy/blob/pgl/src/wrapper/WCKTON.sol).

Details implementation;

There are two actions designed in KTON precompile:

1. Transfer and call (Darwinia native kton -> wkton contract token)

When using this action, we call the precompile directly. The precompile input includes two-part:

```rs
pub struct CallData {
	wkton_address: H160,
	value: U256,
}
```

Once we receive call from outside and parse the `call_data`, we transfer kton from caller to wkton contract firstly, then call ` deposit(address from, uint256 value)` in WCKTON Contract which deployed in advance.

2. Withdraw (wkton contract token -> Darwinia native kton)

We directly call WCKTON Contract `withdraw(bytes32 to, uint wad)`, this method will make a call inside to the KTON precompile withdraw action.

The input includes two-parts also:

```rs
#[derive(Debug, PartialEq, Eq)]
pub struct WithdrawData<T: frame_system::Config> {
	pub to_account_id: AccountId<T>,
	pub kton_value: U256,
}
```
Once we receive this call and parse the withdrawal data. we transfer kton from wkton contract to the `to_account_id`.

### CODE REVIEW

File: frame/evm/precompile/contracts/kton/src/lib.rs [hunk 1]
```diff
@@ -71,31 +66,38 @@ impl<T: Config> Precompile for Kton<T> {
 			.checked_pow(U256::from(9))
 			.unwrap_or(U256::MAX);
 		let action = which_action::<T>(&input)?;
-		let con_caller = T::AddressMapping::into_account_id(context.caller);
+		// let con_caller = T::AddressMapping::into_account_id(context.caller);
```
Comment (songtianyi) [inline frame/evm/precompile/contracts/kton/src/lib.rs:69]: commented code

File: frame/evm/precompile/contracts/kton/src/lib.rs [hunk 2]
```diff
@@ -108,14 +110,39 @@ impl<T: Config> Precompile for Kton<T> {
...
-					&con_caller,
+				// Ensure wkton is a contract
+				ensure!(
+					!crate::Module::<T>::is_contract_code_empty(&context.caller),
+					ExitError::Other("The caller must be wkton contract!".into())
+				);
+				// Ensure context's apparent_value is zero
+				ensure!(
+					context.apparent_value == U256::zero(),
+					ExitError::Other("The value in tx must be zero!".into())
```
Comment (songtianyi) [inline frame/evm/precompile/contracts/kton/src/lib.rs:121]: The apparent value in tx

File: frame/evm/precompile/contracts/kton/src/lib.rs [hunk 3]
```diff
@@ -108,14 +110,39 @@ impl<T: Config> Precompile for Kton<T> {
...
+				);
+				// Ensure caller's balance is enough
+				let caller_kton = T::KtonAccountBasic::account_basic(&context.caller);
+				ensure!(caller_kton.balance >= wd.kton_value, ExitError::OutOfFund);
```
Comment (songtianyi) [inline frame/evm/precompile/contracts/kton/src/lib.rs:125]: `OutOfFund` - we should use the same error name for the insufficient balance situation. I didn't find any other name like this by searching this repo.  
BTW we should group errors in one file/crate i think

File: frame/evm/src/lib.rs [hunk 4]
```diff
@@ -563,6 +563,11 @@ impl<T: Config> Module<T> {
 		account.nonce == U256::zero() && account.balance == U256::zero() && code_len == 0
 	}
 
+	pub fn is_contract_code_empty(address: &H160) -> bool {
```
Comment (songtianyi) [inline frame/evm/src/lib.rs:566]: is_contract_code_empty(address: &H160) -> 

empty(address: &H160) or empty_address(address: &H160)

It seems this function is validating address not the contract

File: frame/evm/precompile/contracts/kton/src/lib.rs (hunk 2 above)
Comment (songtianyi) [reply frame/evm/precompile/contracts/kton/src/lib.rs:121]: I don't know what apparent actually means here, probably i'm wrong

Comment (AsceticBear) [reply frame/evm/precompile/contracts/kton/src/lib.rs:121]: The `apparent` refers to tx value. here refers to context inner field from EVM executor.

File: frame/evm/precompile/contracts/kton/src/lib.rs (hunk 3 above)
Comment (AsceticBear) [reply frame/evm/precompile/contracts/kton/src/lib.rs:125]: The return error and `execute` function signature defined in the EVM crate. Thus you can not find it in this repo.

Comment (AurevoirXavier) [reply frame/evm/precompile/contracts/kton/src/lib.rs:125]: > `OutOfFund` - we should use the same error name for the insufficient balance situation. I didn't find any other name like this by searching this repo.
> BTW we should group errors in one file/crate i think

No you can't.

Each error belong to its `mod pallets;`

File: frame/evm/src/lib.rs (hunk 4 above)
Comment (AsceticBear) [reply frame/evm/src/lib.rs:566]: In the Ethereum system, an address can represent both a user account and a contract account. The difference between them is whether they have contract code or not. Here is represent contract account.

File: frame/evm/precompile/contracts/kton/src/lib.rs (hunk 3 above)
Comment (songtianyi) [reply frame/evm/precompile/contracts/kton/src/lib.rs:125]: Yes, but we should have a name convention for error, event names, etc

File: frame/evm/src/lib.rs [hunk 5]
```diff
@@ -340,6 +336,10 @@ decl_storage! {
 					balance: account.balance,
 					nonce: account.nonce,
 				});
+				T::KtonAccountBasic::mutate_account_basic(&address, Account {
+					balance: account.balance,
+					nonce: account.nonce,
+				});
```
Comment (AsceticBear) [inline frame/evm/src/lib.rs:342]: Give some kton to init evm account for js test.

**Review by HackFisher — APPROVED** [review]

**Pull request merged**